src/main/java/org/elasticsearch/hadoop/mr/EsInputFormat.java

```diff
@@ -0,0 +1,456 @@
+      super(split, job, reporter);
+    }
+
+    @Override
+    public Text createKey() {
+      return new Text();
+    }
+
+    @Override
+    protected Text setCurrentKey(Text hadoopKey, Object object) {
+      if (hadoopKey != null) {
+        hadoopKey.set(object.toString());
+      }
+      return hadoopKey;
+    }
+  }
+
+  /**
+   *
+   */
+  protected static class WritableEsInputRecordReader
+    extends AbstractWritableEsInputRecordReader<Map<Writable, Writable>> {
+
+    private boolean useLinkedMapWritable = true;
+
+    public WritableEsInputRecordReader() {
+      super();
+    }
+
+    public WritableEsInputRecordReader(org.apache.hadoop.mapred.InputSplit split, Configuration job,
+                                       Reporter reporter) {
+      super(split, job, reporter);
+    }
+
+
+    @Override
+    void init(EsInputSplit esSplit, Configuration cfg, Progressable progressable) {
+      useLinkedMapWritable = (!MapWritable.class.getName().equals(HadoopCfgUtils.getMapValueClass(cfg)));
+      super.init(esSplit, cfg, progressable);
+    }
+
+    @Override
+    public Map<Writable, Writable> createValue() {
+      return (useLinkedMapWritable ? new LinkedMapWritable() : new MapWritable());
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    protected Map<Writable, Writable> setCurrentValue(Map<Writable, Writable> hadoopValue, Object object) {
+      if (hadoopValue != null) {
+        hadoopValue.clear();
+        Map<Writable, Writable> val = (Map<Writable, Writable>) object;
+        hadoopValue.putAll(val);
+      }
+      return hadoopValue;
+    }
+  }
+
+  /**
+   *
+   */
+  protected static class JsonWritableEsInputRecordReader extends AbstractWritableEsInputRecordReader<Text> {
+
+    public JsonWritableEsInputRecordReader() {
+      super();
+    }
+
+    public JsonWritableEsInputRecordReader(org.apache.hadoop.mapred.InputSplit split, Configuration job,
+                                           Reporter reporter) {
+      super(split, job, reporter);
+    }
+
+    @Override
+    public Text createValue() {
+      return new Text();
+    }
+
+    @Override
+    protected Text setCurrentValue(Text hadoopValue, Object object) {
+      if (hadoopValue != null) {
+        hadoopValue.set(object.toString());
+      }
+      return hadoopValue;
+    }
+  }
+
+  //
+  // new API - just delegates to the Old API
+  //
+  @Override
+  public List<InputSplit> getSplits(JobContext context) throws IOException {
+    JobConf conf = HadoopCfgUtils.asJobConf(context.getConfiguration());
+    // NOTE: this method expects a ShardInputSplit to be returned (which implements both the old and the new API).
+    return Arrays.asList((InputSplit[]) getSplits(conf, conf.getNumMapTasks()));
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public EsInputRecordReader<K, V> createRecordReader(InputSplit split, TaskAttemptContext context) {
+    return (EsInputRecordReader<K, V>) (isOutputAsJson(context.getConfiguration()) ?
```

Comment:
This is changed. original - https://github.com/elastic/elasticsearch-hadoop/blob/v5.4.0/mr/src/main/java/org/elasticsearch/hadoop/mr/EsInputFormat.java#L392